src/Button/types.ts

```diff
@@ -52,17 +52,17 @@ export type ButtonProps = {
   /**
    * The leading visual which comes before the button content
    */
-  leadingVisual?: React.ElementType | null
+  leadingVisual?: React.ReactNode | React.ElementType | null
```

Comment:
Quickly skimming this:

`ReactNode` = ` string | number | boolean | react.ReactElement<any, string | react.JSXElementConstructor<any>> | Iterable<react.ReactNode> | react.ReactPortal | react.PromiseLikeOfReactNode | null | undefined`

Do we sufficiently handle things like string, number, true, etc here? Do we need additional tests/stories for that, or would we want to limit this a bit more in types?

Reply:
@mattcosta7 my assumption from the parent issues was to support the generic `React.ReactNode` but am happy to constrain it if that would be helpful 👍 I think `React.ReactElement` in particular would satisfy the cases that have been brought up. In particular:

```tsx
import * as React from 'react';

function Component() {
  return 'example';
}

const Memo = React.memo(Component)

const ForwardRef = React.forwardRef(Component)

const Lazy = React.lazy(async () => {
  return {
    default: Component
  };
})

const T1: React.ReactElement = <p>Hello</p>;
const T2: React.ReactElement = <>hi</>;
const T3: React.ReactElement = <Lazy />;
const T4: React.ReactElement = <Memo />;
const T5: React.ReactElement = <ForwardRef />;
```

Reply:
FWIW I think it's a good idea to support all of `ReactNode`, even if it takes a bit more logic, because having to reason whether the JSX you're working with is `ReactNode` vs `ReactElement` can be frustrating. For example, you might want to do something like `leadingVisual={someBoolean && <Icon />}` or `leadingVisual={children}` and it's annoying / less readable to have to wrap those in fragments.

Reply:
Agreed re supporting all of react node, but just making sure we test for all of the edges there

the one you mentioned around specific strings behaving properly, for instance being a good case to handle. 

what do the booelans true/false do, etc.  With tests for all of the edges here I think we should be solid

Reply:
When thinking about the edge cases coming up, I definitely feel more confident for things like `React.ReactElement` than things like strings, numbers, or booleans. I'm not sure what kind of scenarios would exist for these kinds of primitives values so I would be unsure what to test for or document in storybook. If there are examples to use then I could definitely pursue that but without those I'm not sure what to add in order to cover those cases 👀 

From the `boolean` perspective, it would be great to use a ternary over the logical operators. From the `string` or `number` perspective, I assume that `number` would be analogous to the `count` prop and `string` would likely need an element anyways to be styled as a visual.

Reply:
That makes total sense, @mattcosta7. Definitely agreed that usage-wise folks can kind of get away with anything here (like in your example, if a number is not supported well they could return a function that renders a number 😅). This is definitely where I got tripped up on looking for edge cases as, at the end of the day, these aren't supported paths necessarily and we don't really have a way to restrict them.

I could do something like a helper similar to:

```tsx
// This would be the `leadingVisual`, `trailingVisual`, etc type
type Slot = React.ReactNode | React.ElementType;

function renderSlot(slot: Slot): React.ReactNode {
  if (typeof slot === 'string') {
    return slot;
  }

  if (typeof slot === 'function') {
    return createElement(slot);
  }

  if (typeof slot === 'boolean') {
    return null;
  }

  return slot;
}
```

That would make it more obvious what these edge cases may be and test around those? But from the context of a button I'm not sure and would love any advice/suggestions if this is going down a route that isn't really what you were thinking of.

Reply:
Yeah, in that case, just limiting to elementtype might at least be clear enough - we don't necessarily need to handle every possible case internally.  at some point, users gonna do what users gonna do